### PR TITLE
Bderusha/programs import

### PR DESCRIPTION
**Pull request by bderusha**

### CODE REVIEW

File: course_discovery/apps/programs/models.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        help_text=_('The lifecycle status of this Program.'),
+        max_length=24,
+        choices=_choices(
+            constants.ProgramStatus.UNPUBLISHED,
+            constants.ProgramStatus.ACTIVE,
+            constants.ProgramStatus.RETIRED,
+            constants.ProgramStatus.DELETED,
+        ),
+        default=constants.ProgramStatus.UNPUBLISHED,
+        # though this field is not nullable, setting blank=True ensures validators
```
Comment (clintonb) [inline course_discovery/apps/programs/models.py:53]: This is weird. Either we accept the default or we don't use it.

File: course_discovery/apps/programs/models.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        blank=True,
+    )
+
+    marketing_slug = models.CharField(
+        help_text=_('Slug used to generate links to the marketing site'),
+        blank=True,
+        max_length=255
+    )
+
+    def save(self, *a, **kw):
```
Comment (clintonb) [inline course_discovery/apps/programs/models.py:65]: args, kwargs

File: course_discovery/apps/programs/models.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        if self.category == constants.ProgramCategory.XSERIES and self.status == constants.ProgramStatus.ACTIVE:
+            if not self.marketing_slug:
+                raise ValidationError(_(
+                    "Active XSeries Programs must have a valid marketing slug."
+                ))
+
+        return super(Program, self).save(*a, **kw)
+
+    class Meta(object):  # pylint: disable=missing-docstring
+        index_together = ('status', 'category')
```
Comment (clintonb) [inline course_discovery/apps/programs/models.py:79]: Why are these indexed together?

File: course_discovery/apps/programs/models.py (hunk 2 above)
Comment (clintonb) [inline course_discovery/apps/programs/models.py:65]: Here, and everywhere else where they are abbreviated.

File: course_discovery/apps/programs/models.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    def __str__(self):
+        return self.name
+
+
+class ProgramOrganization(TimeStampedModel):
```
Comment (clintonb) [inline course_discovery/apps/programs/models.py:85]: This seems like a premature optimization. Just use a foreign key from a program to an organization.

File: course_discovery/apps/programs/models.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        """
+        if not self.id:  # pylint: disable=no-member
+            # Before creating, check program is not already associated with some organization
+            if ProgramOrganization.objects.filter(program=self.program).exists():
+                raise ValidationError(_('Cannot associate multiple organizations with a program.'))
+
+        return super(ProgramOrganization, self).save(*a, **kw)
+
+
+class CourseRequirement(TimeStampedModel):
```
Comment (clintonb) [inline course_discovery/apps/programs/models.py:110]: This, and the model below, can be replaced with a sorted_m2m from `Program` to `Course`:

``` python
class Program:
  courses = SortedManyToManyField(Course, blank=False)
```

File: course_discovery/settings/base.py [hunk 6]
```diff
@@ -48,6 +48,7 @@
     'course_discovery.apps.catalogs',
     'course_discovery.apps.course_metadata',
     'course_discovery.apps.edx_haystack_extensions',
+    'course_discovery.apps.programs',
```
Comment (clintonb) [inline course_discovery/settings/base.py:51]: I'm not convinced programs should be a separate app. Given that programs require courses, it makes sense to combine the models into a single app.

**Comment by bderusha** [comment]
@clintonb all of your comments are for code that I copied directly from the Programs IDA.  I agree with most of the comments, but am wondering if we think it is better to have matching models which capture some of the complexity within Programs or punt and make these models cleaner since the Programs IDA is already handling this complexity.

Either way I agree that we should move this within the course_metadata app

**Comment by clintonb** [comment]
See #153

**Pull request closed without merging**